### PR TITLE
8304013: Add a fast, non-manual alternative to test/jdk/java/util/zip/ZipFile/TestTooManyEntries

### PR DESCRIPTION
**Pull request by eirbjo**
The TestTooManyEntries test was originally added to validate that ZIP64 files with CEN sizes exceeding what ZipFile  supports are rejected with a ZipException. The test does this by creating a large ZIP file (several gigabytes) with many enties. Because this is resource intensive, the test is currently tagged as manual.  (See #6927)

It would be useful to have a test which asserts the CEN size enforcement, but without the CPU, disk, memory and run time requirements of TestTooManyEntries.  Such a fast test can run non-manual,  without the @requires and manual tags as found in TestTooManyEntries.

This PR adds the EndOfCenValidation test which creates sparse test ZIPs where the CEN is "inflated" such that is matches the size declared in the "End of central directory" records.

While thee sparse files look large, they consume very little disk space on file systems supporting sparse files:

```
16 -rw-r--r--  1   2147483702 Feb  6 18:54 bad-cen-offset.zip
16 -rw-r--r--  1   2147483703 Feb  6 18:54 cen-size-too-large.zip
 8 -rw-r--r--  1          132 Feb  6 18:54 invalid-zen-size.zip
``` 

For good measure, two new test methods are added to excercise the remaining ZipExceptions which ZipFile may throw during validation of the END record .

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304013](https://bugs.openjdk.org/browse/JDK-8304013): Add a fast, non-manual alternative to test/jdk/java/util/zip/ZipFile/TestTooManyEntries

### Reviewers
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Martin Buchholz](https://openjdk.org/census#martin) (@Martin-Buchholz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12231/head:pull/12231` \
`$ git checkout pull/12231`

Update a local copy of the PR: \
`$ git checkout pull/12231` \
`$ git pull https://git.openjdk.org/jdk.git pull/12231/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12231`

View PR using the GUI difftool: \
`$ git pr show -t 12231`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12231.diff">https://git.openjdk.org/jdk/pull/12231.diff</a>

</details>

### CODE REVIEW

**Comment by eirbjo** [comment]
I am tempted to rename this test to "CentralDirectoryTooLarge" or similar. The essential thing tested is the size of the CEN, not the number of entries. Any thoughts?

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java:107]: You could speed this up a bit more by just creating the array representing the original Zip 1 time and modify it each pass.(though it might not speed things up that much)

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java:52]: You might consider more meaningful names for the Paths

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java:47]: I would add a comment which also points to the original test

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java:92]: You could consider using expectThrows and then validation the value of getMessage()

**Review by LanceAndersen — COMMENTED** [review]
Thanks for the updates.

Please see the additional suggestions below

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java]: I've removed these fields now and instead produce the files in the test methods directly, removing the need to name the fields.

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java]: I introduced the templateZip() method which returns a byte[]. While the speed is insignificat here, I think it makes it more explicit that we're producing ZIPs from a template.

File: test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java]: Replaced with expectedExceptions and expectedExceptionsMessageRegExp

Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/CenSizeTooLarge.java]: Now that I finally understand what you meant by this, I've update to use expectThrows and assertEquals :-)

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:113]: You can omit the try block

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:133]: Same comment as above.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:219]: Minor nit, I would use the same comment format for the methods as above and probably add an @throws as needed so javadoc tag checks won't fail

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:223]: Perhaps clarify why you are not writing out any data...

**Review by LanceAndersen — COMMENTED** [review]
I think the changes are are looking good.  Please see the comments below.

thank you again for your efforts here

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:113]: I removed the try/catch block with a quivering lower lip :-)

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Changed to block comment style and added @throws.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Sparse files need to be created as such. The purpose of this delete exists to make this explicit. I moved this closer to the FileChannel.open and added a comment to explain the delete.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: No particular reason, so I opted to output some data instead of explaining.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: I am fine leaving it, but still not convinced it is needed if the setup method calls the cleanup method and should achieve the same result.  Again no big deal.

**Review by LanceAndersen — COMMENTED** [review]
Thank you for the updates.

I am going to kick of a Mach 5 run

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Thanks, I agree it is not technically required in this exact context.

The reason I want to keep it is that someone might do a search  for `StandardOpenOption.SPARSE`, find this code and copy it. I think the delete with the comment makes it explicit that preexisting files will not magically be made sparse after the fact. It makes the code samle self-contained.

**Review by LanceAndersen — APPROVED** [review]
Mach 5 test runs are clean so you are good to go Eirik, thank you again

**Comment by eirbjo** [comment]
Thanks for all your help and careful reviews getting this ball across the net, Lance! Feels good to have one less manual test.

/integrate

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (Martin-Buchholz) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:52]: remove the "the" stutter

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (Martin-Buchholz) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:120]: EOC is not a standard zip file technical term.  Did you mean "END header" ?

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (Martin-Buchholz) [inline test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java:139]: I don't think `@throws` javadoc comments in tests are useful; especially checked ones

**Review by Martin-Buchholz — CHANGES_REQUESTED** [review]

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: "the" stutter has been removed 👍

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: If by "standard" you mean "APPNOTE.TXT", then that uses "end of central directory record" which is a bit long.

The Java implementation seems to use END, like in `ZipFile.ENDHDR`, "END header", etc.

It is probably better to be consistent than correct here, so I've changed the test to use "END header" consistently when referring to the "end of central directory record".

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: I recently added a `@throws` by review comment from Lance. I tend to agree that is not very useful, so I'm removing both ocurrences as per your recommendation.

If Lance wants it back, he can add a review comment saying so. I don't particularly care which professor gets the last word in this matter :-)

**Comment by eirbjo** [comment]
@LanceAndersen Would you like to do a final review after the changes suggested by Martin in his late review?

It is mostly comment changes, and some stray variable renames. I have run the test locally after the last changes, and it passes on my machine.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Please add them back before pushing with `@throws IOException if an error occurs`

Yes I understand they do not add much value but it will reduce noise from IDEs and if/when we add more checks for missing javadoc tags, it will save us from  having to revisit this

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: So, what I would suggest is adding a comment somewhere regarding _end of central directory record_ and note the reference to _ENDHDR_ or _END Header_

I don't think you want to spend too much time here just the wording reference the _end of central directory record_ was clear and where you were abbreviating,  then I would reference ENDHDR

I would be looking at the APP.NOTE  when looking at this test then the ZipFile constant verbiage, but that is just me :-)

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Thanks, I've added `@throws IOException if an error occurs` to all methods throwing `IOException`.

For the record, let me state my personal (rather strong) opinion:

I think this a pretty crazy level of boilerplate noise for a test and that it affects readability negatively. My IDE (IntelliJ) does not complain about this at all. Any IDE which complains should be fixed or configured to not complain. Changing Javadoc to complain about this in a scope like this  would be a poor decision.

I think the PR is ready to be sponsored after this.

Comment (LanceAndersen) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: > I think the PR is ready to be sponsored after this.

Please see my comment regarding the  end of central directory record.

I would prefer to tweak that in a fashion similar to what I indicated as I thought your original version was clearer.  I understand what Martin was indicating of the use of EOC and that could have been addressed by adding  _(EOC)_ or _ENDHDR_ after _end of central directory record_ to make it clearer

I will leave it up to you as to whether you want to make the change but it would be clearer as I think we took a small step backwards.

Either way you should not need a sponsor and should be good to integrated when you are ready

Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: > I hear you but the option is not to use javadoc comments and use block comments :-)

I learned something new today: Javadoc comments and block comments are not the same!

Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: So I changed the references to this ultimate ZIP structure to read like this:

'End of central directory record' (END header)

Example:

```
/**
     * Validate that an 'End of central directory record' (END header)
     * where the value of the CEN size field exceeds the position of
     * the END header is rejected.
```

This way we stick to the 'official' ZIP verbiage but also allow using the shorter 'END header'.

What do you think?

Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: > Either way you should not need a sponsor and should be good to integrated when you are ready

As a non-committer, I still think I need help from a committer to `/sponsor` this change, right?

**Review by LanceAndersen — APPROVED** [review]
Thank you for the updates here Eirik. They look fine.

You are right, you do need a sponsor still and I will do so Monday morning so that Martin has a chance to provide any last input on your comment updates regarding the _End of central directory record_.

**Comment by eirbjo** [comment]
Thanks Lance! Allow me to underline just how much I appreciate that we can have differences of opinion while still being friends and making productive progress. Looking around the world, this is not always a given. I see this as a sign of a healty community.

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (Martin-Buchholz) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: Neither APPNOTE nor JDK doc zip  terminology is great, but a test is not the place to invent something better.

(But I'm surprised no one seems to be using "EOCD")

Comment (eirbjo) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: I think that would be “EOCDR”. (Hides behind the bike shed..)

File: test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java
Comment (Martin-Buchholz) [reply test/jdk/java/util/zip/ZipFile/EndOfCenValidation.java]: > I think this a pretty crazy level of boilerplate noise for a test and that it affects readability negatively. 

I'm with Eirik here.

I've maintained jdk tests for 20 years and have never had to worry about nagging to add javadoc boilerplate.

doclint should restrict itself to checking what's there, not to check for completeness, in tests and private APIs.

And it looks like the jdk makefiles are in fact set up to work this way.  

I see:
./make/Docs.gmk:102:JAVADOC_DISABLED_DOCLINT_WARNINGS := missing

Especially for test methods, the contract of a test is very simple: "do nothing", but possibly throw an informative exception.  If the compiler requires a throws declaration, then I like `throws Throwable`  which evokes this contract.  A javadoc `@throws` declaration is the sort of useless boilerplate that makes people hate Java.

**Review by Martin-Buchholz — APPROVED** [review]

**Pull request closed without merging**
